mongo/integration/server_selection_prose_test.go

```diff
@@ -0,0 +1,151 @@
+// https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection-tests.rst
+func TestServerSelectionProse(t *testing.T) {
+	mt := mtest.New(t, mtest.NewOptions().CreateClient(false))
+	defer mt.Close()
+
+	mtOpts := mtest.NewOptions().Topologies(mtest.Sharded).MinServerVersion("4.9")
+	mt.RunOpts("operationCount-based selection within latency window, with failpoint", mtOpts, func(mt *mtest.T) {
+		_, err := mt.Coll.InsertOne(context.Background(), bson.D{})
+		require.NoError(mt, err, "InsertOne() error")
+
+		hosts := options.Client().ApplyURI(mtest.ClusterURI()).Hosts
+		require.GreaterOrEqualf(mt, len(hosts), 2, "test cluster must have at least 2 mongos hosts")
+
+		// Set a failpoint on a specific mongos host that delays all "find" commands for 500ms. We
+		// need to know which mongos we set the failpoint on for our assertions later.
+		failpointHost := hosts[0]
+		mt.ResetClient(options.Client().
+			SetHosts([]string{failpointHost}))
+		mt.SetFailPoint(mtest.FailPoint{
+			ConfigureFailPoint: "failCommand",
+			Mode: mtest.FailPointMode{
+				Times: 10000,
+			},
+			Data: mtest.FailPointData{
+				FailCommands:    []string{"find"},
+				BlockConnection: true,
+				BlockTimeMS:     500,
+				AppName:         "loadBalancingTest",
+			},
+		})
+
+		// Reset the client with exactly 2 mongos hosts.
+		tpm := newTestPoolMonitor()
+		mt.ResetClient(options.Client().
+			SetAppName("loadBalancingTest").
+			SetHosts(hosts[:2]).
+			SetPoolMonitor(tpm.PoolMonitor))
+
+		// Start 25 goroutines that each run 10 findOne operations. Run 25 goroutines instead of the
+		// 10 that the prose test specifies to reduce intermittent test failures caused by the
+		// random selections not being perfectly even over small numbers of operations.
+		var wg sync.WaitGroup
+		for i := 0; i < 25; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				for i := 0; i < 10; i++ {
+					res := mt.Coll.FindOne(context.Background(), bson.D{})
+					assert.NoError(t, res.Err(), "FindOne() error")
+				}
+			}()
+		}
+		wg.Wait()
+
+		// Get all checkOut events and calculate the number of times each server was selected. The
+		// prose test spec says to use command monitoring events, but those don't include the server
+		// address, so use checkOut events instead.
+		checkOutEvents := tpm.Events(func(evt *event.PoolEvent) bool {
+			return evt.Type == event.GetStarted
+		})
+		counts := make(map[string]int)
+		for _, evt := range checkOutEvents {
+			counts[evt.Address]++
+		}
+		assert.Equal(mt, 2, len(counts), "expected exactly 2 server addresses")
+
+		// Calculate the frequency that the server with the failpoint was selected. Assert that it
+		// was selected less than 25% of the time.
+		frequency := float64(counts[failpointHost]) / float64(len(checkOutEvents))
+		assert.Lessf(mt, frequency, 0.25, "expected failpoint host %q to be selected less than 25%% of the time", failpointHost)
+	})
```

Comment:
`ClearFailPoints` does not seem to track which mongos the failpoint was applied. I think the current behavior may route the `{"configureFailPoint": "failCommand", "mode": "off"}` command to the mongos without the failpoint applied. That may skew results if rerunning this test locally during development.

```suggestion

	        /* Reset client to the failpointHost so ClearFailPoints targets on the correct mongos. */
		mt.ResetClient(options.Client().
			SetHosts([]string{failpointHost}))
	})
```

Reply:
Will update test to make sure the test disables the failpoint.

Reply:
Updated to explicitly clear the failpoint on the selected failpoint host.